pkg/cmd/image/squash.go

```diff
@@ -0,0 +1,442 @@
+		return fmt.Errorf("failed to generate commit image config: %w", err)
+	}
+	commitManifestDesc, _, err := sr.writeContentsForImage(ctx, sr.opt.GOptions.Snapshotter, imageConfig, image.Manifest.Layers[:remainingLayerCount], diffLayerDesc)
+	if err != nil {
+		log.G(ctx).WithError(err).Error("failed to write contents for image")
+		return err
+	}
+	nimg := images.Image{
+		Name:      sr.opt.TargetImageName,
+		Target:    commitManifestDesc,
+		UpdatedAt: time.Now(),
+	}
+	_, err = sr.createSquashImage(ctx, nimg)
+	if err != nil {
+		log.G(ctx).WithError(err).Error("failed to create squash image")
+		return err
+	}
+	cimg := containerd.NewImage(sr.client, nimg)
+	if err := cimg.Unpack(ctx, sr.opt.GOptions.Snapshotter, containerd.WithSnapshotterPlatformCheck()); err != nil {
+		log.G(ctx).WithError(err).Error("failed to unpack squash image")
+		return err
+	}
+	return nil
+}
+
+// applyDiffLayer will apply diff layer content created by createDiff into the snapshotter.
+func (sr *squashRuntime) applyDiffLayer(ctx context.Context, baseImg ocispec.Image, sn snapshots.Snapshotter, layers []ocispec.Descriptor) (
+	diffLayerDesc ocispec.Descriptor, diffID digest.Digest, snapshotID string, retErr error) {
+	var (
+		key    = uniquePart()
+		parent = identity.ChainID(baseImg.RootFS.DiffIDs).String()
+	)
+
+	m, err := sn.Prepare(ctx, key, parent)
+	if err != nil {
+		return diffLayerDesc, diffID, snapshotID, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// NOTE: the snapshotter should be hold by lease. Even
+			// if the cleanup fails, the containerd gc can delete it.
+			if err := sn.Remove(ctx, key); err != nil {
+				log.G(ctx).Warnf("failed to cleanup aborted apply %s: %s", key, err)
+			}
+		}
+	}()
+
+	err = sr.applyLayersToSnapshot(ctx, m, layers)
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to apply layers to snapshot %s", key)
+		return diffLayerDesc, diffID, snapshotID, err
+	}
+	diffLayerDesc, diffID, err = sr.createDiff(ctx, key)
+	if err != nil {
+		return diffLayerDesc, diffID, snapshotID, fmt.Errorf("failed to export layer: %w", err)
+	}
+
+	// commit snapshot
+	snapshotID = identity.ChainID(append(baseImg.RootFS.DiffIDs, diffID)).String()
+
+	if err = sn.Commit(ctx, snapshotID, key); err != nil {
+		if errdefs.IsAlreadyExists(err) {
+			return diffLayerDesc, diffID, snapshotID, nil
+		}
+		return diffLayerDesc, diffID, snapshotID, err
+	}
+	return diffLayerDesc, diffID, snapshotID, nil
+}
+
+func newSquashRuntime(client *containerd.Client, option types.ImageSquashOptions) *squashRuntime {
+	return &squashRuntime{
+		opt:          option,
+		client:       client,
+		namespace:    option.GOptions.Namespace,
+		differ:       client.DiffService(),
+		imageStore:   client.ImageService(),
+		contentStore: client.ContentStore(),
+		snapshotter:  client.SnapshotService(option.GOptions.Snapshotter),
+	}
+}
+
+// copied from github.com/containerd/containerd/rootfs/apply.go
```

Comment:
Should be a permanent URL with the git tag (or the commit hash)

Reply:
The commit hash has been added.